### ISSUE TITLE
UIA: delay when reaching prompt with word navigation

### ISSUE THREAD

**Issue by codeofdusk** [microsoft/terminal#5243]
# Environment

```none
Windows build number: [run `[Environment]::OSVersion` for powershell, or `ver` for cmd] 1909 (18363.752)
Windows Terminal version (if applicable): Open Console (4f8acb4)

Any other software? NVDA, Narrator
```

# Steps to reproduce

1. Open Open Console.
2. With either NVDA or Narrator, move to the start of the console and begin navigation by word.
3. After reaching the word "reserved" (of the copyright text), move to the next word.

# Expected behaviour
The prompt is quickly read.

# Actual behaviour
A lag is observed before the prompt is read. A reduced lag (or none at all) is observed in Windows Terminal.

**Comment by codeofdusk** [comment]
Cc @carlos-zamora.

**Comment by DHowett-MSFT** [comment]
This could be us scanning the entire 9001-line buffer for words. @codeofdusk If you change the buffer size from 120x9001 to 120x50 (in the legacy preferences panel), does the slowdown disappear?

**Comment by codeofdusk** [comment]
> This could be us scanning the entire 9001-line buffer for words. @codeofdusk If you change the buffer size from 120x9001 to 120x50 (in the legacy preferences panel), does the slowdown disappear?

Enabling "use legacy console" in properties completely disables UIA. The "buffer size" is already set to 50, should it be something else?

**Comment by codeofdusk** [comment]
> > This could be us scanning the entire 9001-line buffer for words. @codeofdusk If you change the buffer size from 120x9001 to 120x50 (in the legacy preferences panel), does the slowdown disappear?
> 
> Enabling "use legacy console" in properties completely disables UIA. The "buffer size" is already set to 50, should it be something else?

Changing the buffer size from 50 to 500 has no effect.

**Comment by DHowett-MSFT** [comment]
Ah, sorry, I meant the "traditional" console properties sheet. No need to actually check the "legacy" checkbox. That one switches back to the console host from Windows 8.1 before we renewed our work on it :smile:

Just to confirm, you've got 50 or 500 in the box where I've got 9001:?

![image](https://user-images.githubusercontent.com/14316954/78512503-ecfcb300-7759-11ea-92c4-eaf804ed740c.png)

**Comment by codeofdusk** [comment]
> Ah, sorry, I meant the "traditional" console properties sheet. No need to actually check the "legacy" checkbox. That one switches back to the console host from Windows 8.1 before we renewed our work on it 😄
> 
> Just to confirm, you've got 50 or 500 in the box where I've got 9001:?
> 
> ![image](https://user-images.githubusercontent.com/14316954/78512503-ecfcb300-7759-11ea-92c4-eaf804ed740c.png)

It looks like you've got an image with the alt text "image", what is this?

**Comment by DHowett-MSFT** [comment]
Sorry about that. It is a screenshot of the console properties page. The window is titled, "C:\windows\system32\cmd.exe properties"; there's a tab control with "Layout" selected as the active tab. There are three frames inside the tab control:

* Screen Buffer Size
* Window Size
* Window Position

The Screen Buffer Size frame contains the label "Width" and a disabled spinner box containing the number 111; the label "Height" and an enabled spinner box containing the number 9001, and a ticked checkbox labelled "Wrap text output on resize."

There's another similar frame below Screen Buffer Size which is titled "Window Size". Since it's not the frame I am interested in, its contents are not meaningful right now.

(This is a great learning experience for me, and I apologize for providing inaccessible images. :smile:)

**Comment by codeofdusk** [comment]
> Sorry about that. It is a screenshot of the console properties page. The window is titled, "C:\windows\system32\cmd.exe properties"; there's a tab control with "Layout" selected as the active tab. There are three frames inside the tab control:
> 
> * Screen Buffer Size
> * Window Size
> * Window Position
> 
> The Screen Buffer Size frame contains the label "Width" and a disabled spinner box containing the number 111; the label "Height" and an enabled spinner box containing the number 9001; and a ticked checkbox labelled "Wrap text output on resize."
> 
> There's another similar frame below Screen Buffer Size which is titled "Window Size". Since it's not the frame I am interested in, its contents are not meaningful right now.
> 
> (This is a great learning experience for me, and I apologize for providing inaccessible images. 😄)

Oh I see now. Yeah I have 9001 there. Testing.

**Comment by codeofdusk** [comment]
> > Sorry about that. It is a screenshot of the console properties page. The window is titled, "C:\windows\system32\cmd.exe properties"; there's a tab control with "Layout" selected as the active tab. There are three frames inside the tab control:
> > 
> > * Screen Buffer Size
> > * Window Size
> > * Window Position
> > 
> > The Screen Buffer Size frame contains the label "Width" and a disabled spinner box containing the number 111; the label "Height" and an enabled spinner box containing the number 9001; and a ticked checkbox labelled "Wrap text output on resize."
> > There's another similar frame below Screen Buffer Size which is titled "Window Size". Since it's not the frame I am interested in, its contents are not meaningful right now.
> > (This is a great learning experience for me, and I apologize for providing inaccessible images. 😄)
> 
> Oh I see now. Yeah I have 9001 there. Testing.

Yes slowdown is gone when I change it to 50.

**Comment by DHowett-MSFT** [comment]
This is excellent news. Thank you. @carlos-zamora, I believe we chatted about this. Because conhost doesn't know where the "actual" bottom of the text is, it scans forward almost 120x9001 characters to find out that the buffer consists of nothing but spaces. This is the optimization we talked about with Michael where totally empty rows could be skipped as containing "only spaces".

**Comment by DHowett-MSFT** [comment]
(The reason conhost doesn't know where the actual bottom of the buffer is is because a Win32 application can "address" the entire 9001 lines, even if they are not visible. It was a huge mistake to give win32 applications control over the entire buffer outside of the viewport, but it is a burden we need to live with.)

**Comment by carlos-zamora** [comment]
Makes sense. Aside from that change, I’d be happy to do some more WPR traces on ConHost’s word navigation model. It’d be pretty fun actually haha

**Comment by carlos-zamora** [comment]
Closing as a part of #7000 (particularly closed by #6986)
